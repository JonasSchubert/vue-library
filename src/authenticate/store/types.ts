export const ActionTypes = {
  login: 'login',
  logout: 'logout',
  validateToken: 'validate-token'
};

export const GetterTypes = {
  error: 'error',
  isAdministrator: 'isAdministrator',
  isLoading: 'isLoading',
  isUser: 'isUser',
  loggedIn: 'loggedIn',
  token: 'token',
  userRoles: 'userRoles'
};

export const ModuleType = 'authenticate';

export const MutationTypes = {
  setData: 'setData',
  setError: 'setError',
  setIsLoading: 'setIsLoading'
};

export const RouteTypes = {
  authenticate: 'authenticate'
};
