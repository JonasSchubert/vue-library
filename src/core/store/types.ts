export const GetterTypes = {
  isAnythingLoading: 'isAnythingLoading'
};

export const MutationTypes = {
  decreaseRequests: 'decreaseRequests',
  increaseRequests: 'increaseRequests'
};
