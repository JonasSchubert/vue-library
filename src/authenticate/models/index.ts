export * from './authenticate-module-setup.d';
export * from './authenticate-state.d';
export * from './base-authenticate-state.d';
export * from './login-data.d';
export * from './login-response.d';
export * from './validate-token-data.d';
