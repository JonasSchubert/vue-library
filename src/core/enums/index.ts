export * from './edit-mode.enum';
