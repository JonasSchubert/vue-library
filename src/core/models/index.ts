export * from './base-state.d';
export * from './entity.d';
