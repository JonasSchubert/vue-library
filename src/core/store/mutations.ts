import { MutationTree } from 'vuex';
import { BaseState } from '../models';
import { MutationTypes } from './types';

export const createMutations = <T extends BaseState>(): MutationTree<T> => ({
  [MutationTypes.decreaseRequests]: (state: T): void => { state.requests -= 1; },
  [MutationTypes.increaseRequests]: (state: T): void => { state.requests += 1; }
});
