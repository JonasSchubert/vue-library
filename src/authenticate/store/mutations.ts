import { MutationTree } from 'vuex';
import { MutationTypes } from './types';
import { AuthenticateModuleSetup, AuthenticateState, LoginResponse } from '../models';
import { writeCookie, deleteCookie } from '../../core/cookie';

export const createMutations = <T extends LoginResponse>({ cookieKeyAuthenticationToken, daysTilExpiredAuthenticationCookie }: AuthenticateModuleSetup): MutationTree<AuthenticateState<T>> => ({
  [MutationTypes.setData]: (state: AuthenticateState<T>, { data, saveLoginDataTemporary }: { data: T | undefined; saveLoginDataTemporary: boolean }): void => {
    state.data = data;

    if (data?.token !== undefined && saveLoginDataTemporary) {
      writeCookie(cookieKeyAuthenticationToken, data.token, daysTilExpiredAuthenticationCookie);
    } else if (!!data && data.token === undefined) {
      deleteCookie(cookieKeyAuthenticationToken);
    }
  },
  [MutationTypes.setError]: (state: AuthenticateState<T>, { error }: { error: Error | undefined }): void => {
    state.error = error;
  },
  [MutationTypes.setIsLoading]: (state: AuthenticateState<T>, { isLoading }: { isLoading: boolean }): void => {
    state.isLoading = isLoading;
  }
});
