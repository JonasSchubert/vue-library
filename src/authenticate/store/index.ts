import { Module } from 'vuex';
import { AuthenticateModuleSetup, AuthenticateState, LoginResponse } from '../models';
import { createActions } from './actions';
import { createGetters } from './getters';
import { createMutations } from './mutations';
import { createAuthenticateState } from './state';

export const createAuthenticateModule = <TRootState, T extends LoginResponse>({ cookieKeyAuthenticationToken, customGetters, daysTilExpiredAuthenticationCookie }: AuthenticateModuleSetup<TRootState, T>): Module<AuthenticateState<T>, TRootState> => ({
  namespaced: true,
  actions: createActions(),
  getters: createGetters(customGetters),
  mutations: createMutations({ cookieKeyAuthenticationToken, daysTilExpiredAuthenticationCookie }),
  state: createAuthenticateState()
});
