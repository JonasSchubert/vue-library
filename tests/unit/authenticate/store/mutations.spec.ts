import { UserRole } from '@/authenticate/enums';
import { AuthenticateState, LoginResponse } from '@/authenticate/models';
import { createMutations } from '@/authenticate/store/mutations';
import { MutationTypes } from '@/authenticate/store/types';

describe('authenticate/mutations', () => {
  let state: AuthenticateState<LoginResponse>;

  const mutations = createMutations({ cookieKeyAuthenticationToken: 'unit-test', daysTilExpiredAuthenticationCookie: 1 });

  beforeEach(() => {
    state = {
      data: undefined,
      error: undefined,
      isLoading: false
    };
  });

  describe(`${MutationTypes.setData}`, () => {
    test('should set data', () => {
      // Arrange
      const data: LoginResponse = {
        role: UserRole.User,
        success: true,
        token: 'token'
      };

      // Act
      mutations[MutationTypes.setData](state, { data });

      // Assert
      expect(state.data).toMatchSnapshot({
        role: UserRole.User,
        success: true,
        token: 'token'
      });
    });
  });

  describe(`${MutationTypes.setError}`, () => {
    test('should set error', () => {
      // Arrange
      const error: Error = {
        message: 'message',
        name: 'name'
      };

      // Act
      mutations[MutationTypes.setError](state, { error });

      // Assert
      expect(state.error).toMatchSnapshot({
        message: 'message',
        name: 'name'
      });
    });
  });

  describe(`${MutationTypes.setIsLoading}`, () => {
    test('should set isLoading', () => {
      // Arrange
      const isLoading = true;

      // Act
      mutations[MutationTypes.setIsLoading](state, { isLoading });

      // Assert
      expect(state.isLoading).toBeTruthy();
    });
  });
});
