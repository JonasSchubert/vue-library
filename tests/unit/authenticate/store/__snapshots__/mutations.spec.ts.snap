// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`authenticate/mutations setData should set data 1`] = `
Object {
  "role": 1,
  "success": true,
  "token": "token",
}
`;

exports[`authenticate/mutations setError should set error 1`] = `
Object {
  "message": "message",
  "name": "name",
}
`;
